crates/red_knot_python_semantic/src/symbol.rs

```diff
@@ -593,8 +593,18 @@ fn symbol_by_id<'db>(
                 "__slots__" | "TYPE_CHECKING"
             );
 
-            widen_type_for_undeclared_public_symbol(db, inferred, is_considered_non_modifiable)
-                .into()
+            if scope.is_module_scope(db) && scope.file(db).is_stub(db.upcast()) {
```

Comment:
why only at the module level? Typeshed does things like this as well, and I think there's an argument to be made that the public type for `DeprecatedNameForThing` should also not be unioned with `Unknown` here:

```py
class Foo:
    THING: Final = 42
    DeprecatedNameForThing = THING
```

it feels consistent with the idea behind this PR.

Reply:
I didn't have this restriction initially, but then added it because Carl suggested this as a rule here: https://github.com/astral-sh/ruff/issues/17032#issuecomment-2762217959. What was your reasoning @carljm?

Reply:
I'm going to do the following: I'm going to merge this as-is, and then re-open another PR with that restriction lifted. This should allow us to see the impact in a better way.

Reply:
Sounds good!

Reply:
I do think the same rationale applies.